apps/antidote/src/meta_data_notif_server.erl

```diff
@@ -0,0 +1,336 @@
+%% @doc Meta data notification server. Allows client code to subscribe to the
+%% event when the provided snapshot becomes stable.
+
+-module(meta_data_notif_server).
+
+-include("antidote.hrl").
+-include("meta_data_notif_server.hrl").
+
+-ifdef(EUNIT).
+-include_lib("proper/include/proper.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([start_link/0, start/0,
+         init/0,
+         loop/1,
+         subscribe_to_stable/1,
+         notify/2
+        ]).
+
+-record(state, { mons = [] :: [{pid(), reference()}],
+                 %% NOTE: vector_orddict stores Big -> Small
+                 snaps = vector_orddict:new() ::
+                           vector_orddict:vector_orddict([pid()]),
+                 min_snapshot :: undefined | antidote:snapshot_time()
+               }).
+-type state() :: #state{}.
+
+start_link() ->
+    proc_lib:start_link(?MODULE, init, []).
+
+start() ->
+    proc_lib:start(?MODULE, init, []).
+
+%% @doc API to subscribe to stable snapshot notifications. Notification server
+%% will send notification of the form `#md_stable_snapshot{}` to the calling process.
+%% Same client may subscribe to multiple snapshots at the same time.
+-spec subscribe_to_stable(antidote:snapshot_time()) -> ok.
+subscribe_to_stable(Snapshot) ->
+    whereis(?MODULE) ! {add_snapshot, Snapshot, self()},
+    ok.
+
+%% @doc Meta data server notifies notif server about new stable snapshot.
+-spec notify(stable_time_functions, antidote:snapshot_time()) -> ok.
+notify(stable_time_functions, Snapshot) ->
+    true = maybe_notify_about_snapshot(Snapshot),
+    whereis(?MODULE) ! {stable_snapshot, Snapshot},
+    ok.
+
+%%------------------------------------------------------------------------------
+
+init() ->
+    try erlang:register(?MODULE, self()) of
+        true ->
+            proc_lib:init_ack({ok, self()}),
+            loop(#state{})
+    catch _:_ ->
+            proc_lib:init_ack({error, already_registered})
+    end.
+
+loop(State) ->
+    receive
+        {add_snapshot, Snapshot, Pid} ->
+            ?MODULE:loop(add_snapshot(Snapshot, Pid, State));
+
+        {stable_snapshot, SN}
+          when (State#state.min_snapshot =/= undefined) ->
+            SN1 = wait_for_latest(SN),
+            ?MODULE:loop(notify_subscribers(SN1, State));
+
+        {stable_snapshot, _} ->
+            ?MODULE:loop(State);
+
+        {'DOWN', _MonRef, process, Pid, _} ->
+            ?MODULE:loop(remove_subscriber(Pid, State#state.mons));
+
+        stop ->
+            unregister(?MODULE),
+            exit(normal)
+    end.
+
+%%------------------------------------------------------------------------------
+
+wait_for_latest(SN0) ->
```

Comment:
What is this function supposed to do? I am a bit lost here.

Reply:
It just drains message queue to get the latest stable_snapshot message,
which is send by me meta_data subsystem. In this PR this module is a scatch of the idea how this notification server can work. With current implementation stable_snapshot going to be send to this notif_server every second.